group-coordinator/src/main/java/org/apache/kafka/coordinator/group/GroupStore.java

```diff
@@ -0,0 +1,307 @@
+/*
```

Comment:
@aliehsaeedii Thanks for working on this! I have a few high level comments regarding the `GroupStore`. I want us to agree on what we put and don't put in the store.

* In my mind, the store should hold the state of all the groups. I think that we all agree on this.
* As it holds the state of all groups, I am also tempted by moving all the replay methods to the store as they are the ones updating the state. Is it something that you have considered?
* I think that the store should only have methods to query the state (e.g. get a group, list all groups, etc.). All the other methods (e.g. `validateDeleteGroup`, `maybeDeleteGroup`, `createGroupTombstoneRecords`, etc should not be here in my opinion.
* The API specific methods should resides outside of the store. e.g. `listGroups` is directly linked to the implementation of the admin API, hence it may be better to have it in one of the other managers or in a manager responsible for the admin APIs.
* The store does not need a reference to a MetadataImage.

Reply:
Thanks, @dajac, for the conceptual review.

>I am also tempted by moving all the replay methods to the store

Agree! That makes sense and makes the utests cleaner.

>other methods (e.g. validateDeleteGroup, maybeDeleteGroup, createGroupTombstoneRecords, etc should not be here in my opinion.

Your opinion is valid David. Since other manager classes have a reference to `GroupStore`, I assumed such methods could be inside this class to be accessible by all other manager classes. But I think, later we can have a helper class and these methods could be defined there.

>The API specific methods should resides outside of the store. e.g. listGroups 

`listGroups`: REMOVED

>The store does not need a reference to a MetadataImage.

I removed the `MetadataImage` and the associated methods from this class. Do you mean each manager class should have its own `MetadataImage` instance?